### PR TITLE
Fix ByteArrayIndexInput with nonzero offset

### PR DESCRIPTION
**Pull request by DaveCTurner**
We introduced a bug in #79885 by not accounting for cases where there's a nonzero offset in the underlying byte array. Unfortunately `ESIndexInputTestCase` didn't cover these code paths. This commit extends the test coverage and fixes the bug.

### CODE REVIEW

**Review by iverase — APPROVED** [review]
LGTM

Thanks @DaveCTurner!

File: test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java [hunk 1]
```diff
@@ -60,26 +62,37 @@ protected byte[] randomReadAndSlice(IndexInput indexInput, int length) throws IO
         int readPos = (int) indexInput.getFilePointer();
         byte[] output = new byte[length];
         while (readPos < length) {
-            switch (randomIntBetween(0, 5)) {
-                case 0:
-                    // Read by one byte at a time
-                    output[readPos++] = indexInput.readByte();
+            final var readStrategy = randomFrom(2, 6);
+            switch (readStrategy) {
```
Comment (idegtiarenko) [inline test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java:66]: Above looks suspicious. Should it be `randomIntBetween(0, 9)`?

Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java:66]: Yes, fix incoming in https://github.com/elastic/elasticsearch/pull/93208/files#diff-1313cb767b5d3c0840a46464b34a645015f7dffb344967cf380dba0a5f54d833R65

Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java:66]: and https://github.com/elastic/elasticsearch/pull/93227 fixes this in 8.6 too

**Pull request merged**